core/src/main/java/hudson/security/BasicAuthenticationFilter.java

```diff
@@ -138,7 +142,16 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
             User u = User.get(username);
             ApiTokenProperty t = u.getProperty(ApiTokenProperty.class);
             if (t!=null && t.matchesPassword(password)) {
-                SecurityContextHolder.getContext().setAuthentication(u.impersonate());
+                Authentication auth;
+                try {
+                    auth = u.impersonate();
+                } catch (UsernameNotFoundException x) { // User is invalid in the context
+                    LOGGER.log(WARNING, "API token matched for user "+username+" but the impersonation failed",x);
+                    rsp.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+                    rsp.setHeader("WWW-Authenticate","Basic realm=\"Jenkins user\"");
+                    return;
+                }
+                SecurityContextHolder.getContext().setAuthentication(auth);
```

Comment:
Shouldn't this be `finally`? Otherwise the thread retains the impersonation, and if it's from a thread pool…